### PR TITLE
[New] `sync`/`async`: add `realpath`/`realpathSync` options

### PR DESCRIPTION
**Pull request by SimenB**
Not sure how to write tests for this one. Ideas?

I also went with passing the implementation to `maybeUnwrap` so consumer wouldn't have to care about the option, not sure if it's a good idea or not.

(This is probably the one we want in Jest so we can plug in the hacky `process.bindings` one on node 8...)

### CODE REVIEW

**Comment by ljharb** [comment]
I think i'd write a sync + async test that would ordinarily fail tests, but passes because you overrode `unwrapSymlink`?

File: lib/sync.js [hunk 1]
```diff
@@ -25,19 +25,24 @@ var defaultIsDir = function isDirectory(dir) {
...
-    if (!opts || !opts.preserveSymlinks) {
-        try {
-            return fs.realpathSync(x);
-        } catch (realPathErr) {
-            if (realPathErr.code !== 'ENOENT') {
-                throw realPathErr;
-            }
+var defaultUnwrapSymlink = function unwrapSymlink(x) {
+    try {
+        return fs.realpathSync(x);
```
Comment (ljharb) [inline lib/sync.js:30]: does this need adjusting now that #217 has landed?

File: lib/async.js [hunk 2]
```diff
@@ -25,12 +25,16 @@ var defaultIsDir = function isDirectory(dir, cb) {
     });
 };
 
-var maybeUnwrapSymlink = function maybeUnwrapSymlink(x, opts, cb) {
+var defaultUnwrapSymlink = function unwrapSymlink(x, cb) {
+    fs.realpath(x, function (realPathErr, realPath) {
```
Comment (ljharb) [inline lib/async.js:29]: does this need adjusting now that #217 has landed?

File: lib/sync.js (hunk 1 above)
Comment (SimenB) [reply lib/sync.js:30]: yup!

**Comment by SimenB** [comment]
> I think i'd write a sync + async test that would ordinarily fail tests, but passes because you overrode `unwrapSymlink`?

Sounds good. Can I make it not actually do any symlink stuff but just pretend it does? E.g. add an extra directory after "resolving" the symlink?

**Comment by SimenB** [comment]
Rebased after #217, will write a test after breakfast 🙂

**Comment by SimenB** [comment]
@ljharb pushed some tests, PTAL

File: readme.markdown [hunk 3]
```diff
@@ -117,6 +119,13 @@ default `opts` values:
             return cb(err);
         });
     },
+    unwrapSymlink: function unwrapSymlink(file, cb) {
```
Comment (SimenB) [inline readme.markdown:122]: I'm not too fond of this name - I just copied the name from the code. Should we call it `realpath` and `realpathSync` to mirror `readFile` and `readFileSync` options?

**Comment by SimenB** [comment]
@ljharb thoughts on this?

File: readme.markdown (hunk 3 above)
Comment (ljharb) [reply readme.markdown:122]: i don't have a strong opinion; "realpath" is definitely the typical name for this, and "unwrap" really is just a word that makes sense in my head, not something that actually makes sense :-)

I'll update this to match your suggestions.

**Comment by ljharb** [comment]
Released in v1.17.0.

**Comment by SimenB** [comment]
Thank you!

**Pull request merged**